LightwaveRf Binding

foxy82 opened the pull request.

buildhive commented:
[openhab » openhab #2528](https://buildhive.cloudbees.com/job/openhab/job/openhab/2528/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

teichsta commented:
Thanks @foxy82 for this contribution. Hope to start the Review soon. Thanks, Thomas E.-E.

foxy82 commented:
Hi Thomas,

Any luck on reviewing this?

Neil
On 16 Mar 2015 11:09 pm, "Thomas Eichstädt-Engelen" <
notifications@github.com> wrote:

> Thanks @foxy82 <https://github.com/foxy82> for this contribution. Hope to
> start the Review soon. Thanks, Thomas E.-E.
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/openhab/openhab/pull/2289#issuecomment-81982095>.
>

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/.classpath:
```diff
@@ -0,0 +1,11 @@
+	<classpathentry kind="src" path="src/test/java"/>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER"/>
+	<classpathentry kind="var" path="M2_REPO/org/mockito/mockito-all/1.9.5/mockito-all-1.9.5.jar"/>
```
remove

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/.classpath:
```diff
@@ -0,0 +1,11 @@
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER"/>
+	<classpathentry kind="var" path="M2_REPO/org/mockito/mockito-all/1.9.5/mockito-all-1.9.5.jar"/>
+	<classpathentry kind="var" path="M2_REPO/junit/junit/4.10/junit-4.10.jar"/>
```
remove

foxy82 replied:
IS this the whole .classpath file or just a particular bit?

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,28 @@
+ org.osgi.service.component,
+ org.osgi.service.event,
+ org.slf4j
+Export-Package: org.openhab.binding.lightwaverf
```
remove this export unless other bindings use classes from here

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/AbstractLightwaveRfCommand.java:
```diff
@@ -0,0 +1,35 @@
+import org.openhab.binding.lightwaverf.internal.command.LightwaveRFCommand;
+import org.openhab.binding.lightwaverf.internal.message.LightwaveRfMessageId;
+
+
```
add javadoc (containing @author and @since 1.7.0)

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/AbstractLightwaveRfCommand.java:
```diff
@@ -0,0 +1,35 @@
+package org.openhab.binding.lightwaverf.internal;
```
add license header

foxy82 replied:
I'm doing all the headers/ @author and @since now.

foxy82 commented:
I've just added all the headers/author and since. Do I need to do anything to get them into this pull request? (I've not used Git much)

buildhive commented:
[openhab » openhab #3016](https://buildhive.cloudbees.com/job/openhab/job/openhab/3016/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/AbstractLightwaveRfCommand.java:
```diff
@@ -0,0 +1,47 @@
+import org.openhab.binding.lightwaverf.internal.message.LightwaveRfMessageId;
+
+
+/**
```
also briefly describe the purpose of this class

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/LightwaveRFActivator.java:
```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright (c) 2010-2014, openHAB.org and others.
```
update license header

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/LightwaveRFReceiver.java:
```diff
@@ -0,0 +1,225 @@
+    private static final int DELAY_BETWEEN_RECEIVES_MS = 10;
+    
+    private final CopyOnWriteArrayList<LightwaveRFMessageListener> listeners = new CopyOnWriteArrayList<LightwaveRFMessageListener>();
+    private final ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);
```
not sure about the purpose of a ThreadPool with size 1?

foxy82 replied:
It's a scheduled thread pool so allows us to do "scheduleWithFixedDelay" instead of creating our own thread with a sleep which (in my view is more messy). Also means we don't have to worry about cancelling threads or threads dying as the Threadpool will handle it.

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/LightwaveRFReceiver.java:
```diff
@@ -0,0 +1,225 @@
+
+    /**
+    * Run method, this will listen to the socket and receive messages.
+    * The blocking is stopped when the socket is closed.
```
usually, one would expect a while-loop in the run() method to continuously listen to new datagrams. Please explain why this isn't the case in your implementation (Executor).

foxy82 replied:
You don't need a while loop with a scheduled executor that is set to run with a fixed delay. The executor will automatically re-schedule the thread for you after a fixed delay. This is why I prefer the scheduled thread executor with a single thread over creating a thread with a while loop that you have to be able to interrupt correctly, handle errors etc. The thread pool does this all for you out of the box and makes code much cleaner.

teichsta replied:
thanks for the explanation … could you please add it to the javadoc?

buildhive commented:
[openhab » openhab #3018](https://buildhive.cloudbees.com/job/openhab/job/openhab/3018/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/LightwaveRFSender.java:
```diff
@@ -0,0 +1,178 @@
+    private final DatagramSocket transmitSocket;
+
+    // Boolean to indicate if we are running
+    private boolean running = false;
```
rather use javadoc "/** */" than line comments

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/LightwaveRFSender.java:
```diff
@@ -0,0 +1,178 @@
+    private static final Logger logger = LoggerFactory.getLogger(LightwaveRFSender.class);
+    private static final LightwaveRFCommand STOP_MESSAGE = LightwaveRFCommand.STOP_MESSAGE;
+    private static final Integer ONE = new Integer(1);
+    private static final int MAX_RETRY_ATTEMPS = 5;
```
would it be useful to make the retries configurable?

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/LightwaveRFSender.java:
```diff
@@ -0,0 +1,178 @@
+  * @author Neil Renaud
+  * @since 1.7.0
+ */
+public class LightwaveRFSender implements Runnable, LightwaveRFMessageListener {
```
since the implementations of sender and receiver contains some duplicated i would suggest to introduce a new Superclass

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/LightwaveRfBinding.java:
```diff
@@ -0,0 +1,370 @@
+	protected void internalReceiveCommand(String itemName, Command command) {
+		// the code being executed when a command was sent on the openHAB
+		// event bus goes here. This method is only called if one of the
+		// BindingProviders provide a binding for the given 'itemName'.
```
remove this generic comment

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/LightwaveRfBinding.java:
```diff
@@ -0,0 +1,370 @@
+		// BindingProviders provide a binding for the given 'itemName'.
+		logger.debug("internalReceiveUpdate({},{}) is called!", itemName, newState);
+		internalReceive(itemName, newState);
+	}
```
from my understanding only commands should be processed, so remove this method

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/LightwaveRfBinding.java:
```diff
@@ -0,0 +1,370 @@
+	}
+
+	private void publishUpdate(List<String> itemNames, LightwaveRFCommand message, LightwaveRfBindingProvider provider) {
+		logger.info("Publishing Update {} to {}",message, itemNames);
```
decrease to DEBUG level

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/LightwaveRfGenericBindingProvider.java:
```diff
@@ -0,0 +1,229 @@
+
+
+/**
+ * This class is responsible for parsing the binding configuration.
```
please add some valid configuration samples here

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/LightwaveRfGenericBindingProvider.java:
```diff
@@ -0,0 +1,229 @@
+		//	throw new BindingConfigParseException("item '" + item.getName()
+		//			+ "' is of type '" + item.getClass().getSimpleName()
+		//			+ "', only Switch- and DimmerItems are allowed - please check your *.items configuration");
+		//}
```
does this binding really allow all ItemTypes?

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/LightwaveRfHeatPoller.java:
```diff
@@ -0,0 +1,92 @@
+  * @author Neil Renaud
+  * @since 1.7.0
+ */
+public class LightwaveRfHeatPoller  {
```
yet another class which shares many code with xxxSender and xxxReceiver (Superclass?)

foxy82 replied:
Whilst there are lots of schedulers (one in sender to process the sending queue, one in receiver to process the socket messages, one in HeatPoller to add a delays between polls) they all have distinct jobs. I'm not sure a super class would add much due to their different roles. 

It is sort of akin to saying 
public void run(){
 while(variable) { 
  // do something
 Thread.sleep(1000)
}
should all be factored into a super class.

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/command/LightwaveRfHeatingInfoResponse.java:
```diff
@@ -0,0 +1,235 @@
+	 * 		"nSlot":"06:00",
+	 * 		"prof":5
+	 * }
+	 */  
```
add to classlevel javadoc or remove

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/command/LightwaveRfOnOffCommand.java:
```diff
@@ -0,0 +1,133 @@
+	 * Commands are like: 
+	 *     100,!R2D3F0 (Off)
+	 *     101,!R2D3F1 (On)
+	 */
```
move to classlevel

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/main/java/org/openhab/binding/lightwaverf/internal/command/LightwaveRfSetHeatingTemperatureCommand.java:
```diff
@@ -0,0 +1,89 @@
+public class LightwaveRfSetHeatingTemperatureCommand extends AbstractLightwaveRfCommand implements LightwaveRfRoomMessage { 
+
+	private static final Pattern REG_EXP = Pattern.compile("([0-9]{1,3}),!R([0-9])DhF\\*tP([0-9\\.]{1,5}).*\\s*");
+//	private static final Pattern REG_EXP = Pattern.compile("([0-9]{1,3}),!R([0-9])DhF\\*tP([0-9\\.]{1,4})");
```
remove commented code

teichsta commented on bundles/binding/org.openhab.binding.lightwaverf/src/test/java/org/openhab/binding/lightwaverf/internal/LightwaveRFSenderTest.java:
```diff
@@ -0,0 +1,49 @@
+20:03:07.115 DEBUG o.o.b.l.i.LightwaveRfBinding[:203]- internalReceiveCommand(LivingRoom_CubeLight,OFF) is called!
+20:03:07.129 DEBUG o.o.b.l.i.LightwaveRfBinding[:203]- internalReceiveCommand(LivingRoom_DiningTable,OFF) is called!
+20:03:07.151 DEBUG o.o.b.l.i.LightwaveRfBinding[:203]- internalReceiveCommand(Hallway_GroundFloor,OFF) is called!
+ */
```
remove comments

teichsta commented:
Hi @foxy82,

please find my review comments inline and some general comments below.

* update @since to 1.7.0
* add javadoc for ALL class on class level and (at least) for all public methods
* use the Eclipse Code Formatter for all classes
* move the UnitTests into their own bundle ".test” (see other test bundles as sample)
* add binding to distribution/pom.xml

And finally squash your commits into one once you've incorporated my feedback.

Thanks, Thomas E.-E.

buildhive commented:
[openhab » openhab #3020](https://buildhive.cloudbees.com/job/openhab/job/openhab/3020/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

teichsta commented:
if you could finally squash your commits into one i am ready to merge … Thanks, Thomas E.-E.

teichsta commented:
any chance to get this squashed today?

foxy82 commented:
Hey,

I'll try to do it later today.

Neil
On 23 May 2015 07:33, "Thomas Eichstädt-Engelen" <notifications@github.com>
wrote:

> any chance to get this squashed today?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openhab/openhab/pull/2289#issuecomment-104853583>.
>

teichsta commented:
@foxy82 ?

foxy82 commented:
Hi,

I've incorporated the following: fixed javadoc, fixed headers, removed tests from the binding directory, formatting.

I haven't done the comments for public methods yet or squashed it into one commit.

I need to go out now I'll try to do the squash in about 3 hours. If you need this today I'm not likely to be able to add the comments that will need to be a later update.

buildhive commented:
[openhab » openhab #3060](https://buildhive.cloudbees.com/job/openhab/job/openhab/3060/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

teichsta commented:
ok, thanks! The most important thing would be to squash your commits into one and to create the according wiki page. Best, Thomas E.-E.

foxy82 commented:
I'm trying to squash now but I don't know Git very well. I'm following this: http://blog.steveklabnik.com/posts/2012-11-08-how-to-squash-commits-in-a-github-pull-request

but something must be going wrong as loads of commits are conflicting I've only managed to get to 10 of 121 if it persists at this pace I'm not going to have time to do it.

If there is anything wrong with those instructions let me know.

teichsta commented:
we generally recommend this howto: https://github.com/ginatrapani/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

teichsta commented:
i can squash for you but then i will be author here on github. If you are fine with that we can go that way, too …

foxy82 commented:
Following that - I don't know how to find how many commits I need for the HEAD~4 bit just done 20 commits successfully but now it is asking about merge conflicts in db4o modules. I wonder if I've mucked up my fork when trying to pull in upstream changes.

foxy82 commented:
I've managed to squash some but the next one is just conflicting really badly - I don't know why - all I want is all files in the status they are in the very last commit as a single commit for the whole binding. I don't have time to do any more I'll have to look again if I get some time next week. I don't mind if you are able to squash them and it sets you as the author - or if you know a simple command I can run to just flatten it taking the HEAD as the final copy and not worrying about the merge conflicts in the middle.

buildhive commented:
[openhab » openhab #3063](https://buildhive.cloudbees.com/job/openhab/job/openhab/3063/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[openhab » openhab #3064](https://buildhive.cloudbees.com/job/openhab/job/openhab/3064/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[openhab » openhab #3065](https://buildhive.cloudbees.com/job/openhab/job/openhab/3065/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

teichsta commented:
i've merged this PR manually. Please keep in mind to create the according Wiki page and link to it here. Thanks, Thomas E.-E.
